runtime/src/chpl-init.c

```diff
@@ -416,5 +423,4 @@ extern void chpl_deinitModules(void);
 void chpl_library_finalize(void) {
   chpl_libraryModuleLevelCleanup();
   chpl_deinitModules();
-  chpl_rt_finalize(0);
```

Comment:
I might be missing something, but doesn't this completely disable the runtime finalization? I think we want `chpl_rt_finalize()` to be called, because ` chpl_task_exit()`, `chpl_comm_exit()`, etc. should be invoked at some point, shouldn't they...? I think it's just the final `exit(status);` call in [`chpl_exit_common`](https://github.com/chapel-lang/chapel/blob/57e4a64aa1de4b8ee6648ab39abba9492783be24/runtime/src/chplexit.c#L34) that should be removed.

Reply:
Ah, you're right.  I was thinking that `chpl_exit_all()` just did the barrier + exit.  Thanks for the catch.

Reply:
Note that at least gasnet will terminate program execution during `gasnet_exit` before we even get to our `exit`. I don't know of any way to stop gasnet without terminating execution.

Reply:
@ronawho do you happen to know whether GASNet just happens to do that or whether there's really no other way (without major modifications to the code)? I checked what the MPI conduit does, and there's this line https://github.com/chapel-lang/chapel/blob/00bda24c633adfe1f30940dad2a919929e09a06e/third-party/gasnet/gasnet-src/other/ammpi/ammpi_spmd.c#L378
Perhaps, this could be set to a dummy noop :smiling_imp: 

On a more selfish note, in my use case I call `chpl_library_finalize` from Python, and there the code is compiled with `CHPL_COMM=none`, so removing the `exit(status)` would still help.

Reply:
> gasnet will terminate program execution during gasnet_exit

Oh, I was not aware of that!  I've asked the GASNet team whether there's an alternative.

@ronawho: I assume it's important we call gasnet_exit() before we call chpl_mem_exit() or chpl_topo_exit()?  (otherwise, we could just skip gasnet_exit() and have our normal exit() do that...).  Or...  is it only important for non-gasnet comm layers?  (in which case we could make gasnet's chpl_comm_exit() a no-op, leaning on exit() to call gasnet_exit()...).

Meanwhile, I'll see if I can get the PR into a reasonable shape where we won't exit in the comm=none case...  (or, really, where if chpl_comm_exit() doesn't exit, we won't either).

Reply:
I think only important for other comm layers. For instance ugni free's memory during deinint so we still need the memory layer.

Probably we should split this into a deinint and exit phase, where most current runtime exit's would become the deinint and we'd just have to split gasnet out to only shut down AM handler during deinit and exit later. Note that this can leave other resources utilized (at the very least memory in a fixed heap config, but could be NIC resources, and even CPU resources in the case there's a background thread under the covers) but I think that's the best we could do for now.

And given gasnet isn't currently an important use-case here just getting this working for non-exiting comm layers is fine with me